### ISSUE TITLE
Keeping track of file versions

### ISSUE THREAD

**Issue by ashirwad** [workflowr/workflowr#208]
The data files for the research projects that I work on are kept on Box cloud that supports file versioning. I pull these data files into an R session using functions from the `boxr` package that allows users to communicate with Box using its API. I was wondering if you have any ideas on how I can keep track of the file versions that were used to generate analyses for ensuring reproducibility?

**Comment by jdblischak** [comment]
Hi @ashirwad Thanks for using workflowr!

> I was wondering if you have any ideas on how I can keep track of the file versions that were used to generate analyses for ensuring reproducibility?

I haven't used Box myself, but it looks very convenient. I looked through the [available functions](https://cran.r-project.org/web/packages/boxr/boxr.pdf) of the [boxr](https://github.com/r-box/boxr) package, and I saw that it was possible to specify the exact version of the file when you pull it into R.

Thus my first recommendation is to always specify the version ID (`version_id`) or number (`version_no`) when pulling a data file from Box into R, e.g.

```
myData <- box_search("data.csv") %>%
  box_read_csv(version_no = 1)
```

And then if the data gets updated on Box, you can change the line to `version_no = 2`, and run:

```
wflow_publish("analysis/my-analysis.Rmd", message = "Update results with version number 2 of data.csv")
```

If you'd prefer to not have to manually change the version, then my second recommendation is to record the version directly into the HTML file by running [`box_previous_versions()`](https://r-box.github.io/boxr/reference/box_previous_versions.html):

```
fileID <- box_search("data.csv")
myData <-  box_read_csv(fileID, version_no = 1)
box_previous_versions(fileID)
```

I don't know how the data frame returned by `box_previous_versions()` is sorted, but it should be possible to automatically only show the most recent version (i.e. the one that was pulled into R).

Would either of these suggestions work for you?

**Comment by ashirwad** [comment]
Thanks for the prompt response, @jdblischak!

Based on what I understood from the vignette, we can roll back to only the past Rmd versions of the published html file and not the past html files (please correct me if I am wrong). So, if I use your first recommendation, then it will be pretty obvious to know which version of the file I used in the past because it's hard coded in the Rmd file itself. This can get cumbersome if files are updated too often because it will involve manually updating the file version numbers whenever there is a change in the file.

However, if I use your second recommendation, I will be able to automatically keep track of the file versions in the html file itself, but if I want to go back to past versions, I will have to locate the corresponding html file (if it's currently possible), get the version number and then load the corresponding files, correct? I see there is a bit of a trade-off! 

What's your thoughts on which approach will be less painful?

**Comment by jdblischak** [comment]
> What's your thoughts on which approach will be less painful?

I think the latter approach will be less painful. It requires less manual intervention on your part.

By recording the version of the file in the HTML, you will always know what version was used to create those results. Thus if the results unexpectedly change, e.g. a plot looks strange, then you can look at the previous HTML, copy-paste the file version, explicitly pass it to the box function, and re-run the analysis. If the results are fixed, then you know the change is due to the updated file. If it doesn't change, then you know the difference is due to a change you made in the code (or a change in behavior of a function you use from an external package).

Workflowr makes it easy to browse past versions of the HTML pages (using the past versions tab of the workflowr report at the top of each page). If you want to try using the data as it was months ago, you can navigate to that past version of the HTML to retrieve the file ID that was previously used. 

> Based on what I understood from the vignette, we can roll back to only the past Rmd versions of the published html file and not the past html files (please correct me if I am wrong).

I'm not entirely sure I understand. If you use Git to return to a past version of your entire repository, then both the Rmd and HTML files will be returned to their previous versions. But if you didn't manually specify the box file version, you are correct that rebuilding the HTML would give different results (since it will fetch the latest version, and not the version that had been used).

**Comment by ashirwad** [comment]
Ok, it makes total sense now! Thanks again for creating this wonderful package.  Also, I think it will be helpful if you add a vignette to your `workflowr` website explaining basically what we discussed here. This approach to keeping track of file versions will apply to other cloud services such as Google Drive that also allows users to pull data directly into R similar to Box (see [googledrive](https://googledrive.tidyverse.org/) package for example).

**Comment by jdblischak** [comment]
> I think it will be helpful if you add a vignette to your workflowr website explaining basically what we discussed here. 

I'd love to have a vignette discussing this topic. Since you have direct experience using this cloud-based setup, would you be interested in helping co-write the vignette?

**Comment by ashirwad** [comment]
Yeah, I will be more than happy to help!

**Comment by jdblischak** [comment]
@ashirwad Great! To start, please experiment with the different versioning strategies that we discussed above. That way we'll know how they work in practice. Also, would it be possible for me to view your workflowr website so that I can learn from it?

I'll handle getting a draft of the vignette started.

**Comment by ashirwad** [comment]
@jdblischak, yeah sure, I will start experimenting with the versioning ideas that we discussed. Since I just started using `workflowr` I will need a couple of days to become a `workflowr` ninja! ;) I will keep you posted as to my progress. Also, I am building a private `wokflowr` website for my project (client's requirements), but I will see if I can grant a temporary access for you to get a glimpse of it.

**Comment by jdblischak** [comment]
> Also, I am building a private wokflowr website for my project (client's requirements), but I will see if I can grant a temporary access for you to get a glimpse of it.

Totally understandable. Instead of full access, a few screenshots of the HTML output when you use functions like `box_previous_versions()` would be informative.

**Comment by ashirwad** [comment]
Hey @jdblischak, I had submitted a request to add `box_current_version()`  (https://github.com/r-box/boxr/issues/142) to `boxr` authors some time ago. The request has now been approved and the function is now available on their GitHub branch. I will send you the screenshot that you requested soon. Also, sorry for not keeping you posted.

**Comment by jdblischak** [comment]
@ashirwad Very cool! Thanks for all your hard work making the process of analyzing data files on Box more reproducible for R users!

**Comment by ashirwad** [comment]
@jdblischak, `boxr` package developers have added a `boxr::box_version_number()` function to return the version number of a Box file. This is how I am using it: 
 
![image](https://user-images.githubusercontent.com/25992196/96036366-f7df8100-0e29-11eb-9fe0-cfe970c3135e.png)

`boxr::box_version_number()` prints the message as well as returns the integer output for the version number. I wrap `boxr::box_version_number()` inside `invisible()` because message is enough for keeping track of the file version for reproducibility. Also, for interactive development, I use `cache.extra = boxr::box_version_number(file_id)` to invalidate a cache whenever a newer version of a file is available.
